[21606] Add missing tutorials for AML Dashboard (#17)
eProsima-Private/AML-Tools: docs/rst/tutorials/dashboard_inference.rst

Ensure you have installed the *AML Dashboard* using one of the following methods:

- :ref:`Linux Installation <amldashboard_linux>`
- :ref:`Docker Image Installation <amldashboard_docker>`

For more information, check the :ref:`AML Dashboard Interfaces <amldashboard_interfaces_evaluation>` and :ref:`AML Dashboard Usage <amldashboard_usage_evaluation>` sections.

Running the demo
================

To run the necessary components for evaluating a model and making predictions using the |aml| Dashboard, follow these steps:

Start the backend server
------------------------

1. Navigate to the ``backend`` directory.

.. code-block:: bash

   cd backend

2. Load the |amlip| environment.

.. code-block:: bash

   source /AML-IP/install/setup.bash

3. Start the server:

.. code-block:: bash

   python3 server.py

Start the Inference Node
------------------------

.. warning::

    Before starting the inference node, ensure that there is at least one model and traing set available in the download directory.

1. Load the |amlip| environment.

.. code-block:: bash

   source /AML-IP/install/setup.bash

2. Navigate to the ``backend`` directory.

.. code-block:: bash

   cd backend

> 3. Start one or more inference nodes:

Reviewer: ```suggestion
3. Start an inference node:
```

.. code-block:: bash

   python3 inference.py

Start the AML Dashboard
-----------------------

1. Navigate to the ``frontend/aml_dashboard`` directory.

.. code-block:: bash

   cd frontend/aml_dashboard

2. Start the |aml| Dashboard:

.. code-block:: bash

   npm run dev

3. Access the dashboard at `http://localhost:5173/ <http://localhost:5173/>`__.

Evaluating the Model
====================

To evaluate a model using the |aml| Dashboard, follow these steps:

1. Navigate to the ``Batch Prediction`` tab on the |aml| Dashboard.

2. Click on the ``Update Predictions`` button in the *Algebraic Machine Learning* section to initiate the predictions.

3. The confusion matrix for the model's predictions will be displayed. This matrix helps in assessing the model's performance by showing the accuracy and errors.

>    <video id=myVideo width=100% height=auto autoplay loop controls muted>

Reviewer: ```suggestion
.. raw:: html

   <video id=myVideo width=100% height=auto autoplay loop controls muted>
```

      <source src="../../_static/resources/tutorials/dashboard_batch_prediction.mp4">
      Your browser does not support the video tag.
   </video>

   <script>
      // Set the speed of the video once the page is loaded
      window.onload = function() {
         document.getElementById('myVideo').playbackRate = 1.0; // Set speed to 1.0x
      };
   </script>

Real-Time Predictions
=====================

To make real-time predictions using the trained model with the |aml| Dashboard, follow these steps:

1. Go to the ``Real-time Prediction`` tab on the |aml| Dashboard.  

2. Click on the ``toggle prediction`` button in the *Predict for AML* section to start the real-time prediction process.  

3. In the *webcam* section, toggle the ``activate video`` button to enable the webcam.

4. The prediction made by the model will be displayed in real time with the corresponding label. 

>    <video id=myVideo width=100% height=auto autoplay loop controls muted>

Reviewer: ```suggestion
.. raw:: html

   <video id=myVideo width=100% height=auto autoplay loop controls muted>
```

.. include:: /rst/exports/alias.include
.. include:: /rst/exports/roles.include

.. _tutorials_dashboard_inference:

####################################
Make predictions with the AML Dashboard

> ####################################

Reviewer: ```suggestion
#######################################
Make predictions with the AML Dashboard
#######################################
```

Background
==========

The :term:`AML` Dashboard is a web-based tool that allows users to interact with the |aml| framework.

> This tutorial showcases the process of making predictions based on an |aml| model using the *AML Dashboard*.

Reviewer: ```suggestion
This tutorial showcases the process of making predictions using an |aml| model through the *AML Dashboard*, covering both real-time and batch prediction workflows.
```

> .. figure:: /rst/figures/tutorials/inference.png 

Reviewer: ```suggestion
.. figure:: /rst/figures/tutorials/inference_dashboard.png 
```

For more information, check the :ref:`AML Dashboard Interfaces <amldashboard_interfaces_evaluation>` and :ref:`AML Dashboard Usage <amldashboard_usage_evaluation>` sections.

Running the demo
================

To run the necessary components for evaluating a model and making predictions using the |aml| Dashboard, follow these steps:

Start the backend server
------------------------

1. Navigate to the ``backend`` directory.

.. code-block:: bash

   cd backend

2. Load the |amlip| environment.

.. code-block:: bash

   source /AML-IP/install/setup.bash

3. Start the server:

.. code-block:: bash

   python3 server.py

Start the Inference Node
------------------------

.. warning::

    Before starting the inference node, ensure that there is at least one model and traing set available in the download directory.

1. Load the |amlip| environment.

.. code-block:: bash

   source /AML-IP/install/setup.bash

2. Navigate to the ``backend`` directory.

.. code-block:: bash

   cd backend

3. Start one or more inference nodes:

.. code-block:: bash

   python3 inference.py

.. check if there can be more inference nodes run parallely. 

> Each inference node will wait for job assignments and will collectively distribute the workload when multiple nodes are running. 

Reviewer: Currently, if multiple inference nodes are run in parallel, data will only be received by one node, and inference will be performed by that single node. Although supporting parallel inference nodes can be implemented in the future, for now, I will remove the option to run multiple nodes in parallel. This will be updated once the capability is available.

Start the AML Dashboard
-----------------------

1. Navigate to the ``frontend/aml_dashboard`` directory.

.. code-block:: bash

   cd frontend/aml_dashboard

2. Start the |aml| Dashboard:

.. code-block:: bash

   npm run dev

3. Access the dashboard at `http://localhost:5173/ <http://localhost:5173/>`__.

Evaluating the Model
====================

To evaluate a model using the |aml| Dashboard, follow these steps:

1. Navigate to the ``Batch Prediction`` tab on the |aml| Dashboard.

2. Click on the ``Update Predictions`` button in the *Algebraic Machine Learning* section to initiate the predictions.

> 5. The confusion matrix for the predictions made by the model will be displayed.

Reviewer: ```suggestion
3. The confusion matrix for the model's predictions will be displayed. This matrix helps in assessing the model's performance by showing the accuracy and errors.
```

.. raw:: html

   <video id=myVideo width=100% height=auto autoplay loop controls muted>
      <source src="../../_static/resources/tutorials/dashboard-batch-prediction.mp4">
      Your browser does not support the video tag.
   </video>

   <script>
      // Set the speed of the video once the page is loaded
      window.onload = function() {
         document.getElementById('myVideo').playbackRate = 1.0; // Set speed to 1.0x
      };
   </script>

Real-Time Predictions
=====================

To make real-time predictions using the trained model with the |aml| Dashboard, follow these steps:

> 1. Navigate to the ``Real-time Prediction`` tab on the |aml| Dashboard.

Reviewer: ```suggestion
1. Go to the ``Real-time Prediction`` tab on the |aml| Dashboard.
```

> 2. Check on the ``toggle prediction`` button in the *Predict for AML* section to initiate the real-time prediction process.

Reviewer: ```suggestion
2. Click on the ``toggle prediction`` button in the *Predict for AML* section to start the real-time prediction process.
```

> 

Reviewer: Please, add the activate webcam step.

> 5. The label for the predictions made by the model will be displayed.

Reviewer: ```suggestion
3. The prediction made by the model will be displayed in real time with the corresponding label.
```

> .. raw:: html
> 
>    <video id=myVideo width=100% height=auto autoplay loop controls muted>
> 
>    <video id=myVideo width=100% height=auto autoplay loop controls muted>

Reviewer: ```suggestion
.. raw:: html

   <video id=myVideo width=100% height=auto autoplay loop controls muted>
```
